pkg/controller/mysql/grant/reconciler.go

```diff
@@ -0,0 +1,324 @@
+}
+
+func parseGrant(grant, dbname string) (privileges []string) {
+	matches := grantRegex.FindStringSubmatch(grant)
+	if len(matches) == 3 && matches[2] == dbname {
+		return strings.Split(matches[1], ", ")
+	}
+	return nil
+}
+
+func (c *external) getPrivileges(ctx context.Context, username, dbname string) ([]string, *managed.ExternalObservation, error) {
+	username, host := mysql.SplitUserHost(username)
+	query := fmt.Sprintf("SHOW GRANTS FOR %s@%s", mysql.QuoteValue(username), mysql.QuoteValue(host))
+	rows, err := c.db.Query(ctx, xsql.Query{String: query})
+	if err != nil {
+		return nil, nil, errors.Wrap(err, errCurrentGrant)
+	}
+	defer rows.Close() //nolint:errcheck
+
+	var privileges []string
+	for rows.Next() {
+		var grant string
+		if err := rows.Scan(&grant); err != nil {
+			return nil, nil, errors.Wrap(err, errCurrentGrant)
+		}
+		p := parseGrant(grant, dbname)
+		if p != nil && privileges != nil {
+			// Found more than one grant for this user on this DB
+			return nil, &managed.ExternalObservation{
+				ResourceExists:   true,
+				ResourceUpToDate: false,
+			}, nil
+		}
+		privileges = p
+	}
+	if err := rows.Err(); err != nil {
+		var myErr *mysqldriver.MySQLError
+		if errors.As(err, &myErr) {
+			if myErr.Number == errCodeNoSuchGrant {
+				// The user doesn't (yet) exist and therefore no grants either
+				return nil, &managed.ExternalObservation{ResourceExists: false}, nil
+			}
+		}
+		return nil, nil, errors.Wrap(err, errCurrentGrant)
+	}
+	if privileges == nil {
+		return nil, &managed.ExternalObservation{ResourceExists: false}, nil
+	}
+	return privileges, nil, nil
+}
+
+func (c *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotGrant)
+	}
+
+	username := *cr.Spec.ForProvider.User
+	dbname := *cr.Spec.ForProvider.Database
+
+	privileges := strings.Join(cr.Spec.ForProvider.Privileges, ", ")
+
+	query := createGrantQuery(privileges, dbname, username)
+	if err := c.db.Exec(ctx, xsql.Query{String: query}); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateGrant)
+	}
+	if err := c.db.Exec(ctx, xsql.Query{
+		String: "FLUSH PRIVILEGES",
+	}); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errFlushPriv)
+	}
+	return managed.ExternalCreation{}, nil
+}
+
+func (c *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotGrant)
+	}
+
+	username := *cr.Spec.ForProvider.User
+	dbname := *cr.Spec.ForProvider.Database
+
+	privileges := strings.Join(cr.Spec.ForProvider.Privileges, ", ")
+	username, host := mysql.SplitUserHost(username)
+
+	// Remove current grants
+	query := fmt.Sprintf("REVOKE ALL ON %s.* FROM %s@%s",
+		mysql.QuoteIdentifier(dbname),
+		mysql.QuoteValue(username),
+		mysql.QuoteValue(host),
+	)
```

Comment:
Could this result in applications running as this user being unable to use the database? e.g. For a small period of time between this query being executed and the following create query being executed? Or perhaps even longer if this query succeeds but the create fails? Is it possible to do this in a transaction?

Reply:
Yes, this could lead to applications having no privileges for a short time. I couldn't find a way to `UPDATE` a grant and not removing it first...

Unfortunately, `GRANT` and/or `REVOKE` statements can't be used in transactions since they lead to an implicit commit of said transaction: https://dev.mysql.com/doc/refman/8.0/en/implicit-commit.html

Reply:
It might be worth capturing this information in a comment for future readers.